### PR TITLE
Support remote storage, step3, only for be, add migration_task_v2

### PR DESCRIPTION
**Pull request by pengxiangyu**
# Proposed changes

Issue Number: #7097

## Problem Summary:

1. Add TStorageMigrationReqV2 and EngineStorageMigrationTask to support migration action
2. Change TabletManager::create_tablet() for remote storage
3. Change TabletManager::try_delete_unused_tablet_path() for remote storage

## Checklist(Required)

1. Does it affect the original behavior: (No)
5. Has unit tests been added: (No)
6. Has document been added or modified: (No)
7. Does it need to update dependencies: (No)
8. Are there any changes that cannot be rolled back: (No)

### CODE REVIEW

File: be/src/olap/rowset/beta_rowset.cpp [hunk 1]
```diff
@@ -154,31 +158,38 @@ OLAPStatus BetaRowset::copy_files_to(const std::string& dir) {
     return OLAP_SUCCESS;
 }
 
-OLAPStatus BetaRowset::upload_files_to(const FilePathDesc& dir_desc) {
+OLAPStatus BetaRowset::upload_files_to(const FilePathDesc& dir_desc, RowsetId new_rowset_id, bool delete_src) {
```
Comment (morningman) [inline be/src/olap/rowset/beta_rowset.cpp:161]:
```suggestion
OLAPStatus BetaRowset::upload_files_to(const FilePathDesc& dir_desc, const RowsetId& new_rowset_id, bool delete_src) {
```

File: be/src/olap/rowset/alpha_rowset.cpp [hunk 2]
```diff
@@ -100,7 +100,7 @@ OLAPStatus AlphaRowset::link_files_to(const FilePathDesc& dir_desc, RowsetId new
     return OLAP_SUCCESS;
 }
 
-OLAPStatus AlphaRowset::copy_files_to(const std::string& dir) {
+OLAPStatus AlphaRowset::copy_files_to(const std::string& dir, RowsetId new_rowset_id) {
```
Comment (morningman) [inline be/src/olap/rowset/alpha_rowset.cpp:103]:
```suggestion
OLAPStatus AlphaRowset::copy_files_to(const std::string& dir, const RowsetId& new_rowset_id) {
```

Comment (morningman) [reply be/src/olap/rowset/alpha_rowset.cpp:103]: Need to check all other places

File: be/src/olap/rowset/beta_rowset.cpp [hunk 3]
```diff
@@ -132,19 +132,23 @@ OLAPStatus BetaRowset::link_files_to(const FilePathDesc& dir_desc, RowsetId new_
...
+OLAPStatus BetaRowset::copy_files_to(const std::string& dir, RowsetId new_rowset_id) {
     for (int i = 0; i < num_segments(); ++i) {
-        FilePathDesc dst_path_desc = segment_file_path(dir, rowset_id(), i);
+        FilePathDesc dst_path_desc = segment_file_path(dir, new_rowset_id, i);
         Status status = Env::Default()->path_exists(dst_path_desc.filepath);
         if (status.ok()) {
-            LOG(WARNING) << "file already exist: " << dst_path_desc.filepath;
-            return OLAP_ERR_FILE_ALREADY_EXIST;
+            LOG(WARNING) << "file already exist, delete it: " << dst_path_desc.filepath;
+            status = Env::Default()->delete_file(dst_path_desc.filepath);
```
Comment (morningman) [inline be/src/olap/rowset/beta_rowset.cpp:141]: The previous behavior just return `OLAP_ERR_FILE_ALREADY_EXIST`.
Why you modify this logic and remove dest path here?

File: be/src/olap/rowset/beta_rowset_writer.cpp [hunk 4]
```diff
@@ -141,6 +141,36 @@ OLAPStatus BetaRowsetWriter::add_rowset_for_linked_schema_change(
     return add_rowset(rowset);
 }
 
+OLAPStatus BetaRowsetWriter::add_rowset_for_migration(RowsetSharedPtr rowset) {
+    OLAPStatus res = OLAP_SUCCESS;
+    assert(rowset->rowset_meta()->rowset_type() == BETA_ROWSET);
```
Comment (morningman) [inline be/src/olap/rowset/beta_rowset_writer.cpp:146]: use `CHECK` instead of `assert`

File: be/src/olap/rowset/beta_rowset_writer.cpp [hunk 5]
```diff
@@ -141,6 +141,36 @@ OLAPStatus BetaRowsetWriter::add_rowset_for_linked_schema_change(
...
+    if (!rowset->rowset_path_desc().is_remote() && !_context.path_desc.is_remote()) {
+        res = rowset->copy_files_to(_context.path_desc.filepath, _context.rowset_id);
+        if (res != OLAP_SUCCESS) {
+            LOG(WARNING) << "copy_files failed. src: " << rowset->rowset_path_desc().filepath
+                         << ", dest: " << _context.path_desc.filepath;
+            return res;
+        }
+    } else if (!rowset->rowset_path_desc().is_remote() && _context.path_desc.is_remote()) {
```
Comment (morningman) [inline be/src/olap/rowset/beta_rowset_writer.cpp:154]: is it possible that `rowset->rowset_path_desc().is_remote()` is true here?

File: be/src/olap/storage_engine.cpp [hunk 6]
```diff
@@ -839,10 +849,42 @@ OLAPStatus StorageEngine::_do_sweep(const string& scan_root, const time_t& local
...
+                    if (!env_util::read_file_to_string(Env::Default(), storage_name_path, &buf).ok()) {
+                        LOG(WARNING) << "read storage_name failed: " << storage_name_path;
+                        continue;
+                    }
+                    FilePathDesc remote_path_desc = scan_root_desc;
+                    remote_path_desc.storage_name = buf.ToString();
+                    boost::algorithm::trim(remote_path_desc.storage_name);
+                    std::shared_ptr<StorageBackend> storage_backend = StorageBackendMgr::instance()->
+                            get_storage_backend(remote_path_desc.storage_name);
+                    if (storage_backend != nullptr) {
```
Comment (morningman) [inline be/src/olap/storage_engine.cpp:864]: What if `storage_backend` is null?

File: be/src/olap/tablet_manager.cpp [hunk 7]
```diff
@@ -1013,6 +1030,46 @@ void TabletManager::try_delete_unused_tablet_path(DataDir* data_dir, TTabletId t
     if (Env::Default()->path_exists(schema_hash_path).ok()) {
         LOG(INFO) << "start to move tablet to trash. tablet_path = " << schema_hash_path;
         FilePathDesc segment_desc(schema_hash_path);
+        string remote_file_param_path = schema_hash_path + REMOTE_FILE_PARAM;
+        if (data_dir->is_remote() && FileUtils::check_exist(remote_file_param_path)) {
+            // it means you must remove remote file for this segment first
+            faststring json_buf;
+            Status s = env_util::read_file_to_string(Env::Default(), remote_file_param_path, &json_buf);
```
Comment (morningman) [inline be/src/olap/tablet_manager.cpp:1037]: better give an example of this json string

File: be/src/olap/task/engine_alter_tablet_task.cpp [hunk 8]
```diff
@@ -35,6 +35,7 @@ EngineAlterTabletTask::EngineAlterTabletTask(const TAlterTabletReqV2& request)
 }
 
 OLAPStatus EngineAlterTabletTask::execute() {
+    DorisMetrics::instance()->storage_migrate_v2_requests_total->increment(1);
```
Comment (morningman) [inline be/src/olap/task/engine_alter_tablet_task.cpp:38]: Why adding `storage_migrate_v2_requests_total` here?

File: be/src/olap/storage_migration_v2.cpp [hunk 9]
```diff
@@ -0,0 +1,463 @@
...
+using std::vector;
+
+namespace doris {
+
+DEFINE_GAUGE_METRIC_PROTOTYPE_5ARG(storage_migration_mem_consumption, MetricUnit::BYTES, "",
+                                   mem_consumption, Labels({{"type", "storage_migration"}}));
+
+
+StorageMigrationV2Handler::StorageMigrationV2Handler()
+        : _mem_tracker(MemTracker::create_tracker(
```
[the remark this answers is not in the document]
Comment (pengxiangyu) [reply be/src/olap/storage_migration_v2.cpp:60]: REGISTER_HOOK_METRIC里面用到了

File: be/src/olap/storage_engine.cpp (hunk 6 above)
Comment (pengxiangyu) [reply be/src/olap/storage_engine.cpp:866]: is storage_backend is null, remote storage is invalid, data need not to be deleted. Only local files need to be deleted.

**Review by morningman — APPROVED** [review]
LGTM

**Pull request merged**